### PR TITLE
Add TileDB vectorstore implementation

### PR DESCRIPTION
**Pull request by NikolaosPapailiou**
- **Description:** Add [TileDB](https://tiledb.com) vectorstore implementation. TileDB offers ANN search capabilities using the [TileDB-Vector-Search](https://github.com/TileDB-Inc/TileDB-Vector-Search) module. It provides serverless execution of ANN queries and storage of vector indexes both on local disk and cloud object stores (i.e. AWS S3). More details in:
  - [Why TileDB as a Vector Database](https://tiledb.com/blog/why-tiledb-as-a-vector-database)
  - [TileDB 101: Vector Search](https://tiledb.com/blog/tiledb-101-vector-search)
- **Twitter handle:** @tiledb

### CODE REVIEW

File: libs/langchain/pyproject.toml [hunk 1]
```diff
@@ -300,6 +301,7 @@ all = [
     "momento",
     "singlestoredb",
     "tigrisdb",
+    "tiledb-vector-search",
```
Comment (baskaryan) [inline libs/langchain/pyproject.toml:304]: only need to add to extra_testing extra for unit tests, so no need here. and preference to not add any dependencies to pyproject which aren't need for testing, so no changes needed to pyproject at all

File: libs/langchain/langchain/vectorstores/tiledb.py [hunk 2]
```diff
@@ -0,0 +1,771 @@
...
+            embeddings = OpenAIEmbeddings()
+            db = TileDB(embeddings, index_uri, metric)
+
+    """
+
+    def __init__(
+        self,
+        embedding: Embeddings,
+        index_uri: str,
+        metric: str,
```
Comment (baskaryan) [inline libs/langchain/langchain/vectorstores/tiledb.py:65]: nit: can we add `*` to signature after metric

File: libs/langchain/langchain/vectorstores/tiledb.py [hunk 3]
```diff
@@ -0,0 +1,771 @@
...
+    @property
+    def embeddings(self) -> Optional[Embeddings]:
+        # TODO: Accept embedding object directly
+        return None
+
+    def process_index_results(
+        self,
+        ids: List[int],
+        scores: List[float],
+        k: int = 4,
```
Comment (baskaryan) [inline libs/langchain/langchain/vectorstores/tiledb.py:118]: can we add `*` to all non-inherited methods to separate args and kwargs

File: libs/langchain/langchain/vectorstores/tiledb.py [hunk 4]
```diff
@@ -0,0 +1,771 @@
...
+                uri=self.vector_index_uri,
+                config=self.config,
+                timestamp=self.timestamp,
+                **kwargs,
+            )
+
+    @property
+    def embeddings(self) -> Optional[Embeddings]:
+        # TODO: Accept embedding object directly
+        return None
```
Comment (baskaryan) [inline libs/langchain/langchain/vectorstores/tiledb.py:112]: should return self.embedding

**Review by baskaryan — COMMENTED** [review]
this is great, thanks @NikolaosPapailiou! few comments

File: libs/langchain/pyproject.toml (hunk 1 above)
Comment (NikolaosPapailiou) [reply libs/langchain/pyproject.toml]: Applied.

File: libs/langchain/langchain/vectorstores/tiledb.py (hunk 2 above)
Comment (NikolaosPapailiou) [reply libs/langchain/langchain/vectorstores/tiledb.py:65]: Applied.

File: libs/langchain/langchain/vectorstores/tiledb.py (hunk 3 above)
Comment (NikolaosPapailiou) [reply libs/langchain/langchain/vectorstores/tiledb.py:119]: Applied.

File: libs/langchain/langchain/vectorstores/tiledb.py (hunk 4 above)
Comment (NikolaosPapailiou) [reply libs/langchain/langchain/vectorstores/tiledb.py]: Applied.

**Comment by NikolaosPapailiou** [comment]
Thanks for the quick review @baskaryan! Let me know if you need any more changes

File: libs/langchain/langchain/vectorstores/tiledb.py (hunk 3 above)
Comment (baskaryan) [reply libs/langchain/langchain/vectorstores/tiledb.py:119]: ah sorry should clarify, we can (unfortunately) only do this for methods that aren't already defined on the  base `VectorStore`. so we can't do it for similarity_search_by_vector, similarity_search, max_marginal_relevance_search_by_vector, max_marginal_relevance_search,  add_texts, from_texts

Comment (NikolaosPapailiou) [reply libs/langchain/langchain/vectorstores/tiledb.py:119]: Thanks for the clarification! 
Applied.

**Pull request merged**